### PR TITLE
Iss 329 rule blocked corerules 286

### PR DESCRIPTION
**Pull request by htu**
Instruction for testing this PR:

1. Run the test cases:
`python -m pytest -s tests/unit/test_operations/test_get_codelist_attributes.py`

2. Download the following files:
[rule-286.json.txt](https://github.com/cdisc-org/cdisc-rules-engine/files/11761116/rule-286.json.txt)
[ts-test-dataset-F.json.txt](https://github.com/cdisc-org/cdisc-rules-engine/files/11736766/ts-test-dataset-F.json.txt)
[ts-test-dataset-T.json.txt](https://github.com/cdisc-org/cdisc-rules-engine/files/11736769/ts-test-dataset-T.json.txt)

3. Create a folder called ./dist/issue329 and save the download file there.

4.  Rename the files by removing the .txt

5. Run rule tests using the downloaded files (make sure that you use path relative on your computer and rename the files properly):
`python core.py test -s sdtmig -v 3-4 -dp ./dist/issue329/ts-test-dataset-T.json -r ./dist/issue329/rule-286.json -ct "sdtmct-2020-03-27"`
`python core.py test -s sdtmig -v 3-4 -dp ./dist/issue329/ts-test-dataset-F.json -r ./dist/issue329/rule-286.json -ct "sdtmct-2020-03-27"`

### CODE REVIEW

**Comment by nhaydel** [comment]
@htu can you provide the test data and test rule in the PR description

File: resources/schema/CORE-base.json
Comment (gerrycampion) [inline resources/schema/CORE-base.json:1]: Can you apply prettier formatting to this file? It should eliminate most of these diffs.

File: cdisc_rules_engine/models/operation_params.py [hunk 1]
```diff
@@ -20,6 +20,9 @@ class OperationParams:
     datasets: List[dict]
     standard: str
     standard_version: str
+    ct_package: list = None
```
Comment (nhaydel) [inline cdisc_rules_engine/models/operation_params.py:23]: Should this be plural: `ct_packages`

File: cdisc_rules_engine/operations/get_ct_attribute_values.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+        # 4.0 merge the two datasets by CC
+        # -------------------------------------------------------------------
+        cc_key = ct_data[cc].to_list()
+        ct_list = ct_cache[(ct_cache[cc].isin(cc_key))]
+        n = self.params.dataframe.shape[0]
+        result = pd.Series([ct_list[cv].values[0] for _ in range(n)])
+        return result
+
+    def _get_ct_from_cache(self, ct_key: str, ct_val: str):
```
Comment (nhaydel) [inline cdisc_rules_engine/operations/get_ct_attribute_values.py:43]: Now that we are trying to add docstrings to all functions, would you mind adding them to these two functions? You can follow this guide on how they should be written: https://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings

File: cdisc_rules_engine/utilities/rule_processor.py [hunk 3]
```diff
@@ -249,6 +249,10 @@ def perform_rule_operations(
                 standard_version=standard_version,
                 meddra_path=kwargs.get("meddra_path"),
                 whodrug_path=kwargs.get("whodrug_path"),
+                ct_version=operation.get("version"),
+                ct_attribute=operation.get("attribute"),
+                ct_package=kwargs.get("ct_package"),
+                # ct_package=kwargs.get("codelist_term_maps"),
```
Comment (nhaydel) [inline cdisc_rules_engine/utilities/rule_processor.py:255]: Why was this removed? I don't think it should be

File: cdisc_rules_engine/operations/operations_factory.py [hunk 4]
```diff
@@ -68,6 +71,7 @@ class OperationsFactory(FactoryInterface):
         "extract_metadata": ExtractMetadata,
         "get_column_order_from_dataset": DatasetColumnOrder,
         "get_column_order_from_library": LibraryColumnOrder,
+        "get_ct_attribute_values": CTAttributeValues,
```
Comment (nhaydel) [inline cdisc_rules_engine/operations/operations_factory.py:74]: Maybe we could just call it "get_codelist_attributes"

**Review by nhaydel — CHANGES_REQUESTED** [review]

File: cdisc_rules_engine/operations/operations_factory.py (hunk 4 above)
Comment (htu) [reply cdisc_rules_engine/operations/operations_factory.py:74]: Renamed it to get_codelist_attributes.

File: cdisc_rules_engine/operations/get_ct_attribute_values.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        ct_packages = self.params.ct_package
+        ct_term_maps = (
+            []
+            if ct_packages is None
+            else [self.cache.get(package) or {} for package in ct_packages]
```
[the remark this answers is not in the document]
Comment (htu) [reply cdisc_rules_engine/operations/get_ct_attribute_values.py:48]: Yes. Where can I get api_key for https://github.com/cdisc-org/cdisc-rules-engine/blob/main/cdisc_rules_engine/services/cdisc_library_service.py to get all the packages?

File: cdisc_rules_engine/operations/get_ct_attribute_values.py (hunk 2 above)
Comment (htu) [reply cdisc_rules_engine/operations/get_ct_attribute_values.py:43]: I did and will sync them to the repo.

File: cdisc_rules_engine/operations/get_ct_attribute_values.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nhaydel) [reply cdisc_rules_engine/operations/get_ct_attribute_values.py:48]: https://wiki.cdisc.org/display/LIBSUPRT/Getting+Started%3A+Access+to+CDISC+Library+API+using+API+Key+Authentication

File: cdisc_rules_engine/models/operation_params.py (hunk 1 above)
Comment (htu) [reply cdisc_rules_engine/models/operation_params.py]: Renamed singular to plural.

File: cdisc_rules_engine/operations/get_ct_attribute_values.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (htu) [reply cdisc_rules_engine/operations/get_ct_attribute_values.py]: Added logic to get codelist from library.

File: cdisc_rules_engine/operations/get_ct_attribute_values.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        )
+
+        # convert codelist to dataframe
+        ct_result = {ct_key: [], ct_val: []}
+
+        for item in ct_term_maps:
+            ct_result[ct_key].append(item["package"])
+            codes = sorted(set(code for code in item.keys() if code != "package"))
```
[the remark this answers is not in the document]
Comment (htu) [reply cdisc_rules_engine/operations/get_ct_attribute_values.py]: It was just for helping me to manually verify codes. Do they need to be in the original order?

File: cdisc_rules_engine/utilities/rule_processor.py (hunk 3 above)
Comment (htu) [reply cdisc_rules_engine/utilities/rule_processor.py:255]: I don't need the codelist_term_maps to be pre-processed. I process them in the operation operator.

File: resources/schema/CORE-base.json
Comment (htu) [reply resources/schema/CORE-base.json:1]: The env you helped setup applied the prettier format automatically. Thanks!

File: cdisc_rules_engine/operations/get_ct_attribute_values.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (htu) [reply cdisc_rules_engine/operations/get_ct_attribute_values.py]: I tested with my api_key and it worked.

File: cdisc_rules_engine/operations/get_ct_attribute_values.py (hunk 2 above)
Comment (htu) [reply cdisc_rules_engine/operations/get_ct_attribute_values.py]: Docstrings are added.

File: cdisc_rules_engine/utilities/rule_processor.py (hunk 3 above)
Comment (nhaydel) [reply cdisc_rules_engine/utilities/rule_processor.py:255]: That may be true for this operation, but I don't think it is true fro other operators

File: cdisc_rules_engine/operations/get_codelist_attributes.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+        # 3.0 get dataset records
+        # -------------------------------------------------------------------
+        ct_data = self._get_ct_from_dataset(ct_key=cc, ct_val=cv)
+
+        # 4.0 merge the two datasets by CC
+        # -------------------------------------------------------------------
+        cc_key = ct_data[cc].to_list()
+        ct_list = ct_cache[(ct_cache[cc].isin(cc_key))]
+        n = self.params.dataframe.shape[0]
```
Comment (nhaydel) [inline cdisc_rules_engine/operations/get_codelist_attributes.py:59]: We should have a more descriptive variable name

File: cdisc_rules_engine/operations/get_codelist_attributes.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+        ct_cd = ct_attr
+        if ct_attr == "Term CCODE":
+            ct_cd = "TSVALCD"
+        sel_cols = [c1, c2, ct_cd, cc]
+
+        # get dataframe from dataset records
+        df = self.params.dataframe
+
+        # add CT_PACKAGE column
+        df[cc] = df.apply(
```
Comment (nhaydel) [inline cdisc_rules_engine/operations/get_codelist_attributes.py:137]: Why do we need to add a column to this dataset, and why is it specific to sdtmct?

**Review by nhaydel — CHANGES_REQUESTED** [review]

File: cdisc_rules_engine/operations/get_codelist_attributes.py (hunk 7 above)
Comment (htu) [reply cdisc_rules_engine/operations/get_codelist_attributes.py:59]: Used longer name and added comments for variables

File: cdisc_rules_engine/operations/get_codelist_attributes.py (hunk 8 above)
Comment (htu) [reply cdisc_rules_engine/operations/get_codelist_attributes.py:137]: This is to form controlled term package name. It is not specific to SDTM. It is to handle two ways ("CDISC", "CDISC CT") of controlled term packages are named in the different version of SDTM.

File: cdisc_rules_engine/operations/get_codelist_attributes.py [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+            retrieved from the dataset.
+        """
+        ct_packages = self.params.ct_packages
+        ct_attr = self.params.ct_attribute  # attribute variable specified in rule
+        c1 = self.params.target  # target variable specified in rule
+        c2 = self.params.ct_version  # controlled term version
+        cc = ct_key  # controlled term variable as key
+        ct_cd = ct_attr
+        if ct_attr == "Term CCODE":
+            ct_cd = "TSVALCD"
```
Comment (nhaydel) [inline cdisc_rules_engine/operations/get_codelist_attributes.py:131]: Does this operation need to be TS specific? And can the variables above be named more clearly

File: cdisc_rules_engine/utilities/rule_processor.py (hunk 3 above)
Comment (nhaydel) [reply cdisc_rules_engine/utilities/rule_processor.py:255]: I still think this should be added back, unless this arg is not used by other operations

**Review by nhaydel — CHANGES_REQUESTED** [review]

File: cdisc_rules_engine/utilities/rule_processor.py (hunk 3 above)
Comment (htu) [reply cdisc_rules_engine/utilities/rule_processor.py:255]: Removed the comment so we will have two variables related to controlled terminologies.

File: cdisc_rules_engine/operations/get_codelist_attributes.py (hunk 9 above)
Comment (htu) [reply cdisc_rules_engine/operations/get_codelist_attributes.py:131]: I think that it is trial summary specific if they specify the attribute is "Term CCODE" otherwise, rule author can use whatever variable as attribute column.

File: cdisc_rules_engine/operations/get_ct_attribute_values.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (htu) [reply cdisc_rules_engine/operations/get_ct_attribute_values.py]: Since there is no specific reason that we have to keep it in the original order, I kept the sorting in the code.

[the remark this answers is not in the document]
Comment (nhaydel) [reply cdisc_rules_engine/operations/get_ct_attribute_values.py]: It is an unnecessary step and should be removed

[the remark this answers is not in the document]
Comment (htu) [reply cdisc_rules_engine/operations/get_ct_attribute_values.py]: It is removed and the test cases are changed accordingly.

**Comment by nhaydel** [comment]
@htu this is approved already
